flw/src/main/java/org/motechproject/nms/flw/service/impl/FrontLineWorkerServiceImpl.java

```diff
@@ -17,11 +30,87 @@
+        this.settingsFacade = settingsFacade;
+
+        schedulePurgeOfOldFrontLineWorkers();
+    }
+
+    /**
+     * Use the MOTECH scheduler to setup a repeating job
+     * The job will start today at the time stored in flw.purge_invalid_flw_start_time in flw.properties
+     * It will repeat every flw.purge_invalid_flw_ms_interval milliseconds (default value is a day)
+     */
+    private void schedulePurgeOfOldFrontLineWorkers() {
+        //Calculate today's fire time
+        DateTimeFormatter fmt = DateTimeFormat.forPattern("H:m");
+        String timeProp = settingsFacade.getProperty(FLW_PURGE_TIME);
+        DateTime time = fmt.parseDateTime(timeProp);
+        DateTime today = DateTime.now()                     // This means today's date...
+                .withHourOfDay(time.getHourOfDay())         // ...at the hour...
+                .withMinuteOfHour(time.getMinuteOfHour())   // ...and minute specified in imi.properties
+                .withSecondOfMinute(0)
+                .withMillisOfSecond(0);
+
+        //Millisecond interval between events
+        String intervalProp = settingsFacade.getProperty(FLW_PURGE_MS_INTERVAL);
+        Long msInterval = Long.parseLong(intervalProp);
+
+        LOGGER.debug(String.format("The %s message will be sent every %sms starting at %s",
+                                    FLW_PURGE_EVENT_SUBJECT, msInterval.toString(), today.toString()));
+
+        //Schedule repeating job
+        MotechEvent event = new MotechEvent(FLW_PURGE_EVENT_SUBJECT);
+        RepeatingSchedulableJob job = new RepeatingSchedulableJob(
+                event,          //MOTECH event
+                today.toDate(), //startTime
+                null,           //endTime, null means no end time
+                null,           //repeatCount, null means infinity
+                msInterval,     //repeatIntervalInMilliseconds
+                true);          //ignorePastFiresAtStart
+
+        schedulerService.safeScheduleRepeatingJob(job);
+    }
+
+    @MotechListener(subjects = { FLW_PURGE_EVENT_SUBJECT })
+    public void purgeOldInvalidFLWs(MotechEvent event) {
+        int weeksToKeepInvalidFLWs = Integer.parseInt(settingsFacade.getProperty(WEEKS_TO_KEEP_INVALID_FLWS));
+        final FrontLineWorkerStatus status = FrontLineWorkerStatus.INVALID;
+        final DateTime cutoff = DateTime.now().minusWeeks(weeksToKeepInvalidFLWs).withTimeAtStartOfDay();
+
+        @SuppressWarnings("unchecked")
+        QueryExecution<Long> queryExecution = new QueryExecution<Long>() {
+            @Override
+            public Long execute(Query query, InstanceSecurityRestriction restriction) {
+
+                query.setFilter("status == invalid && invalidationDate < cutoff");
+                query.declareParameters("org.motechproject.nms.flw.domain.FrontLineWorkerStatus invalid, org.joda.time.DateTime cutoff");
+
+                return query.deletePersistentAll(status, cutoff);
```

Comment:
We should have one on status for sure.  If there is one on status and
invalidationDate then it would work for both cases.  I'll add it.

On Sat, May 23, 2015 at 5:41 PM, Frank Huster <notifications@github.com>
wrote:

> In
> flw/src/main/java/org/motechproject/nms/flw/service/impl/FrontLineWorkerServiceImpl.java
> <https://github.com/motech-implementations/mim/pull/275#discussion_r30949312>
> :
>
> > +
> > +    @MotechListener(subjects = { FLW_PURGE_EVENT_SUBJECT })
> > +    public void purgeOldInvalidFLWs(MotechEvent event) {
> > +        int weeksToKeepInvalidFLWs = Integer.parseInt(settingsFacade.getProperty(WEEKS_TO_KEEP_INVALID_FLWS));
> > +        final FrontLineWorkerStatus status = FrontLineWorkerStatus.INVALID;
> > +        final DateTime cutoff = DateTime.now().minusWeeks(weeksToKeepInvalidFLWs).withTimeAtStartOfDay();
> > +
> > +        @SuppressWarnings("unchecked")
> > +        QueryExecution<Long> queryExecution = new QueryExecution<Long>() {
> > +            @Override
> > +            public Long execute(Query query, InstanceSecurityRestriction restriction) {
> > +
> > +                query.setFilter("status == invalid && invalidationDate < cutoff");
> > +                query.declareParameters("org.motechproject.nms.flw.domain.FrontLineWorkerStatus invalid, org.joda.time.DateTime cutoff");
> > +
> > +                return query.deletePersistentAll(status, cutoff);
>
> Now is probably not the time (Sprint 4 might be) but wouldn't you like to
> have an index on invalidationDate & status?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/motech-implementations/mim/pull/275/files#r30949312>.
>